pkg/controllers/static/deprovisioning/controller.go

```diff
@@ -0,0 +1,179 @@
+
+	scaleDownErrs := make([]error, len(candidates))
+	actualTerminatedCount := int64(0)
+	// Terminate selected NodeClaims
+	workqueue.ParallelizeUntil(ctx, len(candidates), len(candidates), func(i int) {
+		candidate := candidates[i]
+
+		if err := retry.OnError(retry.DefaultBackoff, func(err error) bool { return client.IgnoreNotFound(err) != nil }, func() error {
+			return c.kubeClient.Delete(ctx, candidate.NodeClaim)
+		}); err != nil && client.IgnoreNotFound(err) != nil {
+			log.FromContext(ctx).Error(err, "failed to delete NodeClaim", "NodeClaim", klog.KObj(candidate.NodeClaim))
+			scaleDownErrs[i] = err
+			return
+		}
+
+		atomic.AddInt64(&actualTerminatedCount, 1)
+
+		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, helper.TerminationReason)...)
+		metrics.NodeClaimsDisruptedTotal.Inc(map[string]string{
+			metrics.ReasonLabel:       pretty.ToSnakeCase(helper.TerminationReason),
+			metrics.NodePoolLabel:     candidate.NodeClaim.Labels[v1.NodePoolLabelKey],
+			metrics.CapacityTypeLabel: candidate.NodeClaim.Labels[v1.CapacityTypeLabelKey],
+		})
+		c.cluster.MarkForDeletion(candidate.NodeClaim.Status.ProviderID)
+	})
+
+	log.FromContext(ctx).WithValues("NodePool", klog.KObj(np)).
+		Info("scaled down static nodepool", "current", currentNodeClaims, "desired", desiredReplicas, "terminated", actualTerminatedCount)
+
+	if actualTerminatedCount != nodeClaimsToTerminate {
+		return reconcile.Result{RequeueAfter: time.Second}, serrors.Wrap(
+			errors.NewAggregate(scaleDownErrs),
+			"reason", helper.TerminationReason,
+		)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("deprovisioning.static").
+		For(&v1.NodePool{}, builder.WithPredicates(nodepoolutils.IsManagedPredicateFuncs(c.cloudProvider), predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return true
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldNP := e.ObjectOld.(*v1.NodePool)
+				newNP := e.ObjectNew.(*v1.NodePool)
+				return HasNodePoolReplicaOrStatusChanged(oldNP, newNP)
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return false
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return false
+			},
+		})).
+		Watches(&v1.NodeClaim{}, nodeutils.NodeClaimEventHandler(c.kubeClient), builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return true
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return e.ObjectOld.GetDeletionTimestamp().IsZero() && !e.ObjectNew.GetDeletionTimestamp().IsZero()
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return true
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return false
+			},
+		})).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
```

Comment:
Any rationale behind the choice of `10` for MaxConcurrentReconciles?

Reply:
Since we have parallelism to provision/deprovision multiple NodeClaims per reconcile loop, starting with a lower number to not overwhelm the API server. We can increase later as needed.